### PR TITLE
Inverse Kinematics for Animations

### PR DESCRIPTION
**Pull request by BMagnu**
This PR introduces an animation segment type used to combine multiple submodels into a chain which can be controlled by inverse kinematics.
Allows for each joint to have a specified rotation constraint, as well as to target a specific orientation with the end effector.
Implements the FABRIK algorithm with additions for constraints and end effector rotation.
The hinge constraint _could_ cause certain cases where FABRIK doesn't properly converge, but these'd need to be investigated in the wild when found.

As a side note: Apart from the outstanding translation module by Goober (for which animation support already exists), this is the final thing on my ToDo for animations. So unless someone has more feature requests, the overhaul can be considered basically complete with this PR.

### CODE REVIEW

**Comment by BMagnu** [comment]
Converted to draft to quickly add moveable support

**Review by Baezon — APPROVED** [review]
Looks like a pretty good implementation of the FABRIK algorithm to me. FABRIK is mentioned, but should there perhaps be a link to the paper?

File: code/math/ik_solver.cpp [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+				// Convert new local rotation and global rotation of tha parent's parent to new parent global rotation
+				vm_matrix_x_matrix(&parent.calculatedRot, &localRot, &parentRot);
+				
+				// Find new child position from new parent global rotation, child offset and parent position
+				vm_vec_unrotate(&child.calculatedPos, &child.submodel->offset, &parent.calculatedRot);
+				child.calculatedPos += parent.calculatedPos;
+			}
+		}
+		
+		//After positioning the end effector, it now needs to be oriented correctly. If we have a target orientation use it, otherwise assume it's parents rotation. Then constrain.
```
Comment (Goober5000) [inline code/math/ik_solver.cpp:110]: it's -> its

**Review by Goober5000 — COMMENTED** [review]
That's about it

**Review by Goober5000 — APPROVED** [review]

**Pull request merged**